surfacers/prometheus/prometheus.go

```diff
@@ -373,11 +376,58 @@ func (ps *PromSurfacer) record(em *metrics.EventMetrics) {
 
 // writeData writes metrics data on w io.Writer
 func (ps *PromSurfacer) writeData(w io.Writer) {
+	nowTime := time.Now()
 	for _, name := range ps.metricNames {
+		if metricKeys, ok := ps.getExpiredMetrics(name, nowTime); ok {
```

Comment:
I think add more story about this situation. There are file called `cloudprober.cfg`. That file is set below code.

``` txt
probe {
  name: "cloudprober-0"
  type: PING
  targets {
    file_targets {
      file_path: "/cfg/config-0.json"
      re_eval_sec: 60
    }
  }
}
```
``` txt
{"resource": [{"name": "www.google.com"}, {"name": "XXX"}, {"name": "YYY"}]}
``` 
`/cfg/config-0.json` is set like above file.

After some time, I changed the file because we wanted to ping check targets.
I will update this file below json file.

``` txt
{"resource": [{"name": "ZZZ"}, {"name": "XXX"}, {"name": "YYY"}]}
``` 

Here comes the problem.

Surfacer does not know that the information is missing. That's why Promethues keeps giving out old data every time it takes data.

Reply:
I expected that old target data is removed and Prometheus don't scrap this data. 
But cloudprober continuously provided older data. 

Prometheus occurs error when scraped data is older than 10 min. [link](https://github.com/prometheus/prometheus/issues/6554#issuecomment-571876561)

Promethues had to keep scraping the old data.
As a result, Prometheus did not work properly because there were more data that were not normal than normal data.

I will explain about the example above and show you the previous situation and this pr goal.

### As-be
``` txt
#TYPE total counter
total{ptype="http",probe="cloudprober-0",dst="www.google.com"} 16 ......... (old data)
total{ptype="http",probe="cloudprober-0",dst="XXX"} 16 1617887882590
total{ptype="http",probe="cloudprober-0",dst="YYY"} 16 1617887882590
total{ptype="http",probe="cloudprober-0",dst="ZZZ"} 16 1617887882590
```

### To-be
``` txt
total{ptype="http",probe="cloudprober-0",dst="XXX"} 16 1617887882590
total{ptype="http",probe="cloudprober-0",dst="YYY"} 16 1617887882590
total{ptype="http",probe="cloudprober-0",dst="ZZZ"} 16 1617887882590
```

Reply:
This remaining data is scraped by Promethues. And Prometheus make error that named by `out of bound`. 
Because this remaining data is very old data or will be old data , and Promethues just read without any filtering.

Reply:
So this (targets going away and not updating anymore) is a very common phenomenon. Based on the issue you linked above and prometheus source code, this error should occur only in two scenarios:

1. Data is too far into the future (by 10m).
2. Prometheus has scraped a newer value of a metric and it scrapes an older value after that.

Both of these scenarios make sense to me. A metric simply not getting updated should not cause any problem. It's possible that I am getting the 2nd scenario wrong. I'll try to reproduce this.

I think this is the relevant code in prometheus:
https://github.com/prometheus/prometheus/blob/80545bfb2eb8f9deeedc442130f7c4dc34525d8d/tsdb/head.go#L1258

Reply:
I don't think it's common for the data to not be updated.

I think that keeping unmanaged data stored constantly. It is same about leaveing unused memory.

> It's possible that I am getting the 2nd scenario wrong. I'll try to reproduce this.

I didn't understand exactly what it was.

Are you saying about working on modifying Prometheus?

Or saying that solving the problem through new PR than this?

Reply:
> It's possible that I am getting the 2nd scenario wrong. I'll try to reproduce this.

I didn't understand exactly what it was.

Are you saying about working on modifying Prometheus?

Or saying that solving the problem through new PR than this?